halo2_proofs/src/plonk/prover.rs

```diff
@@ -353,16 +357,26 @@ where
                 );
 
                 // Add blinding factors to advice columns
-                for advice_values in &mut advice_values {
+                for (column_index, advice_values) in &mut advice_values.iter_mut().enumerate() {
                     for cell in &mut advice_values[unusable_rows_start..] {
                         *cell = Scheme::Scalar::random(&mut rng);
+                        if witness.unblinded_advice.contains(&column_index) {
+                            *cell = Blind::default().0;
+                        } else {
+                            *cell = Scheme::Scalar::random(&mut rng);
+                        }
                     }
                 }
 
                 // Compute commitments to advice column polynomials
-                let blinds: Vec<_> = advice_values
-                    .iter()
-                    .map(|_| Blind(Scheme::Scalar::random(&mut rng)))
+                let blinds: Vec<_> = (0..meta.num_advice_columns)
+                    .map(|i| {
+                        if witness.unblinded_advice.contains(&i) {
```

Comment:
Also need to use `column_indices` here.

```suggestion
                let blinds: Vec<_> = column_indices
                    .iter()
                    .map(|i| {
                        if witness.unblinded_advice.contains(i) {
```